cassovary-core/src/main/scala/com/twitter/cassovary/graph/MemoryMappedDirectedGraph.scala

```diff
@@ -0,0 +1,153 @@
+
+  private def outDegree(id: Int): Int = ((outboundOffset(id + 1) - outboundOffset(id)) / 4).toInt
+
+  private def inboundOffset(id: Int): Long = data.getLong(8L + 8L * (nodeCount + 1) + 8L * id)
+
+  private def inDegree(id: Int): Int = ((inboundOffset(id + 1) - inboundOffset(id)) / 4).toInt
+
+  /* Only created when needed (there is no array of these stored). */
+  private class MemoryMappedDirectedNode(override val id: Int) extends Node {
+    val nodeOutboundOffset = outboundOffset(id)
+    val nodeInboundOffset = inboundOffset(id)
+    override def outboundNodes(): Seq[Int] = new IndexedSeq[Int] {
+      val length: Int = outDegree(id)
+      def apply(i: Int): Int =  data.getInt(nodeOutboundOffset + 4L * i)
+    }
+    override def inboundNodes(): Seq[Int] = new IndexedSeq[Int] {
+      val length: Int = inDegree(id)
+      def apply(i: Int): Int =  data.getInt(nodeInboundOffset + 4L * i)
+    }
+  }
+
+  override def getNodeById(id: Int): Option[Node] =
+    if(0 <= id && id < nodeCount) {
+      Some(new MemoryMappedDirectedNode(id))
+    } else {
+      None
+    }
+
+  override def iterator: Iterator[Node] = (0 to nodeCount).iterator flatMap (i => getNodeById(i))
+
+  override def edgeCount: Long = outboundOffset(nodeCount) - outboundOffset(0)
+
+  // Uncomment in future if maxNodeId becomes a method:
+  // override def maxNodeId = nodeCount - 1
+
+  override val storedGraphDir = StoredGraphDir.BothInOut
+}
+
+object MemoryMappedDirectedGraph {
+  /** Writes the given graph to the given file (overwriting it if it exists) in the current binary format.
+   */
+  def graphToFile(graph: DirectedGraph[Node], file: File): Unit = {
+    val n = graph.maxNodeId + 1 // includes both 0 and maxNodeId as ids
+    val out = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(file)))
+    out.writeInt(0)
+    out.writeInt(n)
+    //The outneighbor data starts after the initial 8 bytes, n+1 Longs for outneighbors and n+1 Longs for in-neighbors
+    var outboundOffset = 8L + 8L * (n + 1) * 2
+    for (i <- 0 until n) {
+      out.writeLong(outboundOffset)
+      outboundOffset += 4 * (graph.getNodeById(i) map (_.outboundCount)).getOrElse(0)
+    }
+    out.writeLong(outboundOffset) // Needed to compute outdegree of node n-1
+
+    // The inbound data starts immediately after the outbound data
+    var inboundOffset = outboundOffset
+    for (i <- 0 until n) {
+      out.writeLong(inboundOffset)
+      inboundOffset += 4 * (graph.getNodeById(i) map (_.inboundCount)).getOrElse(0)
+    }
+    out.writeLong(inboundOffset) // Needed to compute indegree of node n-1
+
+    for (i <- 0 until n) {
+      for (v <- (graph.getNodeById(i) map (_.outboundNodes())).getOrElse(Nil)) {
+        out.writeInt(v)
+      }
+    }
+    for (i <- 0 until n) {
+      for (v <- (graph.getNodeById(i) map (_.inboundNodes())).getOrElse(Nil)) {
+        out.writeInt(v)
+      }
+    }
+    out.close()
+  }
+}
+
+object MemoryMappedDirectedGraphBenchmark {
```

Comment:
this and the main() can not be in cassovary-core as it is a library. Ideally should be in cassovary-benchmarks. I suggest you move to cassovary-examples for now and I can refactor it later.